File: backend/src/app/service_layer/schemas/vendor.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator

from app.adapters.db.models.enums import BankGroup, VendorCategory, VendorStatus


class FinancialScore(BaseModel):
    """5C profiling. Skala 0-100."""

    character: int | None = Field(None, ge=0, le=100)
    capacity: int | None = Field(None, ge=0, le=100)
    capital: int | None = Field(None, ge=0, le=100)
    collateral: int | None = Field(None, ge=0, le=100)
    condition: int | None = Field(None, ge=0, le=100)


class VendorDocuments(BaseModel):
    sptTahunan: str | None = None
    neraca: str | None = None
    anggaranDasar: str | None = None
    izinPerusahaan: str | None = None
    rekening: str | None = None


class VendorBase(BaseModel):
    npwp: str = Field(..., min_length=15, max_length=25)
    company_name: str = Field(..., min_length=3, max_length=255)
    company_type: str = Field(..., max_length=20)
    director_name: str = Field(..., min_length=3, max_length=255)
    # Nullable supaya vendor lama (sebelum fitur ini ada) tetap valid saat dibaca.
    category: VendorCategory | None = None
    city: str = Field(..., max_length=100)
    address: str = Field(..., max_length=500)
    email: EmailStr
    phone: str = Field(..., min_length=8, max_length=30)
    bank: BankGroup
    bank_name: str = Field(..., max_length=100)
    bank_account_no: str | None = Field(None, max_length=50)

    @field_validator("npwp")
    @classmethod
    def normalize_npwp(cls, v: str) -> str:
        return v.strip()


class VendorCreate(VendorBase):
    # Wajib diisi saat registrasi baru, walau nullable di VendorBase/VendorRead.
    category: VendorCategory


class VendorUpdate(BaseModel):
    company_name: str | None = None
    company_type: str | None = None
    director_name: str | None = None
    category: VendorCategory | None = None
    city: str | None = None
    address: str | None = None
    email: EmailStr | None = None
    phone: str | None = None
    bank: BankGroup | None = None
    bank_name: str | None = None
    bank_account_no: str | None = None
    documents: VendorDocuments | None = None
    financial_score: FinancialScore | None = None


class VendorVerificationUpdate(BaseModel):
    """Menggerakkan vendor di alur verifikasi 8 langkah."""

    verification_step: int = Field(..., ge=0, le=8)
    status: VendorStatus
    note: str | None = None


class VendorRead(VendorBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: VendorStatus
    verification_step: int
    documents: dict = {}
    financial_score: dict | None = None
    performance_rating: float | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/vendor.ts
export const VENDOR_STATUS = ['pending', 'need-verification', 'verified', 'rejected'] as const
export type VendorStatus = (typeof VENDOR_STATUS)[number]

export const BANK_GROUP = ['Mandiri', 'Bank Lainnya'] as const
export type BankGroup = (typeof BANK_GROUP)[number]

export const VENDOR_CATEGORY = ['Vendor Operasional', 'Vendor Pengadaan', 'Cleaning Service'] as const
export type VendorCategory = (typeof VENDOR_CATEGORY)[number]

/** 5C profiling, skala 0-100 */
export interface FinancialScore {
  character?: number | null
  capacity?: number | null
  capital?: number | null
  collateral?: number | null
  condition?: number | null
}

export interface VendorDocuments {
  sptTahunan?: string | null
  neraca?: string | null
  anggaranDasar?: string | null
  izinPerusahaan?: string | null
  rekening?: string | null
}

export interface Vendor {
  id: string
  npwp: string
  company_name: string
  company_type: string
  director_name: string
  category: VendorCategory | null
  city: string
  address: string
  email: string
  phone: string
  bank: BankGroup
  bank_name: string
  bank_account_no: string | null
  status: VendorStatus
  /** 0-8, mengikuti alur verifikasi 8 langkah */
  verification_step: number
  documents: VendorDocuments
  financial_score: FinancialScore | null
  performance_rating: number | null
  created_at: string
  updated_at: string
}

export interface VendorListQuery {
  page?: number
  size?: number
  search?: string
  status?: VendorStatus
  bank?: BankGroup
}

export const DOCUMENT_FIELDS: { key: keyof VendorDocuments; label: string; required: boolean }[] = [
  { key: 'sptTahunan', label: 'SPT Tahunan', required: true },
  { key: 'neraca', label: 'Neraca Keuangan', required: true },
  { key: 'anggaranDasar', label: 'Anggaran Dasar / Akta Pendirian', required: true },
  { key: 'izinPerusahaan', label: 'Izin Perusahaan (SIUP/NIB)', required: true },
  { key: 'rekening', label: 'Informasi Rekening Bank', required: false },
]

export const VERIFICATION_STEPS = [
  'Upload Dokumen',
  'Verifikasi Logistik',
  'Verifikasi Keuangan',
  'Verifikasi Legal',
  'Upload Form Aplikasi Vendor',
  'Verifikasi Form Aplikasi Vendor',
  'Verifikasi Struktural',
  'DRI',
] as const
